Orb v3

DeNeutoy opened the pull request:
## Description

Orb-v3 - coming in the next few days! I've also included the script I used to do the geo_opt analysis (mostly based on the one in matbench, but isolated to a single script), because I know that was slightly in the air based on #230. To check, I ran a couple of other models using their predictions in figshare (showing the utility of getting people to to this!):

| stol | structure_rmsd vs_dft | n_sym_ops_mae | symmetry decrease | symmetry match | symmetry increase | n_structures | model |
|---:|-:|---:|---:|---:|---:|---:|:---|
| 1.e-2 | 0.07313 | 1.809 | 0.05561 | 0.8144 | 0.1231 | 2.570e5 | MACE MPA 0 |
| 1.e-5 | 0.06078 | 4.163 | 0.4739 | 0.3636 | 0.1188 | 2.570e5 | esen-30m-oam |
| 1.e-2 | 0.06078 | 2.102 | 0.1616 | 0.7103 | 0.1085 | 2.570e5 | esen-30m-oam |
| 1.e-5 | 0.06391 | 2.033 | 0.04393 | 0.7057 | 0.2453 | 2.570e5 | sevennet-mf-ompa |
| 1.e-2 | 0.06391 | 1.705 | 0.04667 | 0.8181 | 0.1280 | 2.570e5 | sevennet-mf-ompa |
| 1.e-5 | 0.07481 | 10.17 | 0.8694 | 0.1304 | 0.0001440 | 2.570e5 | Orb-v3-con-inf-mpa |
| 1.e-2 | 0.07481 | 2.609 | 0.1592 | 0.7281 | 0.1012 | 2.570e5 | Orb-v3-con-inf-mpa |


We are still in the process of adding the exact models to `orb-models` but we'll make a release over the weekend and then I will update this PR.

## Checklist

Please check the following items before submitting your PR:

- [x] I created a new folder and YAML metadata file `models/<arch_name>/<model_variant>.yml` for my submission. `arch_name` is the name of the architecture and `model_variant.yml` includes things like author details, training set names and important hyperparameters.
- [x] I added the my new model as a new attribute on the [`Model.<arch_name>` enum](https://github.com/janosh/matbench-discovery/blob/57d0d0c8a14cd317/matbench_discovery/enums.py#L274) in `enums.py`.
- [x] I uploaded the energy/force/stress model prediction file for the WBM test set to Figshare or another cloud storage service (`<yyyy-mm-dd>-<model_variant>-preds.csv.gz`).
- [x] I uploaded the model-relaxed structures file to Figshare or another cloud storage service in [JSON lines format](https://jsonlines.org) (`<yyyy-mm-dd>-wbm-IS2RE-FIRE.jsonl.gz`). JSON Lines allows fast loading of small numbers of structures with `pandas.read_json(lines=True, nrows=100)` for inspection.
- [x] I uploaded the phonon predictions to Figshare or another cloud storage service (`<yyyy-mm-dd>-kappa-103-FIRE-<values-of-dist|fmax|symprec>.gz`).
- [x] I included the urls to the Figshare files in the YAML metadata file (`models/<arch_name>/<model_variant>.yml`). If not using Figshare I have included the urls to the cloud storage service in the description of the PR.
- [x] I included the test script (`test_<arch_name>_<task>.py` for `task` in `discovery`, `kappa`, `diatomics`) that generated the prediction files.

## Additional Information (Optional)

- [ ] I included a training script (`train_<arch_name>.py`) if I trained a model specifically for this benchmark.
- [ ] I included a `readme.md` with additional details about my model.

DeNeutoy commented:
@janosh / @CompRhys this is ready now - I am still seeing some error from the tests regarding the csv predictions - I can include them here, but I think it's not being fetched from the aws bucket - is it preferable to have them stored remotely? 

I'll send a follow up PR when the paper is out to update a couple of the yaml sections, shouldn't be too long (but the models are available as of now in `orb-models` on main).

CompRhys commented:
> @janosh / @CompRhys this is ready now - I am still seeing some error from the tests regarding the csv predictions - I can include them here, but I think it's not being fetched from the aws bucket - is it preferable to have them stored remotely?

@janosh has typically been mirroring things to figshare so they can't be deleted and break things and so I think that will fix the test. I will take a quick look but nothing to fix on your side I'd guess, i'll also get the linting issue which is that the license in the v3 yml needs a dash rather than a space to fit the schema

> I'll send a follow up PR when the paper is out to update a couple of the yaml sections, shouldn't be too long (but the models are available as of now in `orb-models` on main).

Sounds good.

janosh commented:
@DeNeutoy ORB v3 looks like an excellent model! 👍 i'm curious what's the inference speed on this one compared to the v2 model?

fyi i converted the ORB-relaxed structures to JSON Lines format before uploading to figshare. it's the new format for all `wbm-geo-opt` files since it allows loading a subset of structures for quick debug runs. was useful for #230

<details><summary>JSON Lines conversion</summary>

```
"""Convert model geometry optimizations in standard JSON format to faster-to-partially-
load JSON Lines format.
"""

# %%
import pandas as pd
from pymatviz.enums import Key

from matbench_discovery.enums import Model

df_model = pd.read_json(Model.orb_v3.geo_opt_path).set_index(Key.mat_id)

df_model.reset_index().to_json(Model.orb_v3.geo_opt_path, orient="records", lines=True)

jsonl_path = Model.orb_v3.geo_opt_path.replace(".json.gz", ".jsonl.gz")
df_lines = pd.read_json(jsonl_path, lines=True).set_index(Key.mat_id)

pd.testing.assert_frame_equal(df_model, df_lines)

```
</details>

could you upload the `test_kappa.py` script that generated `orb-v3-con-inf-mpa-20250404-disp0.03-float64-kappa.json.gz`. also, if you have them could you upload the kappa results for `disp=0.01` and `disp=0.05`? i'm hoping to display multiple displacement values in the future similar to `symprec` `1e-2` and `1e-5`. after that, this is good to go

DeNeutoy commented:
@janosh thanks! Orb-v3 is faster than v2 despite being conservative - it is shallower and wider. We also have some direct models which are available, but I haven't sent PRs for them because the number of them may clutter the leaderboard. We'll have a full analysis of speed in the paper which should be out soon - this is something that might be good to add to the leaderboard in the future (although I understand why it's not there atm, as it's very difficult to measure), as some of the models do not have very forgiving memory/speed profiles. 

Relatedly, it would be good to resolve (or temporarily hide the columns, or something) for RMSD before merging, as the tables online are incorrect.

`disp=0.01` and `disp=0.05` en route, and i'll add the script!

janosh commented:
> Relatedly, it would be good to resolve (or temporarily hide the columns, or something) for RMSD before merging, as the tables online are incorrect.

yeah, i was hoping to have the RMSD fixed two weeks ago but i encountered several delays. i think i can resolve it this week. if not, i'll add a warning message to the site until fixed.

> we'll have a full analysis of speed in the paper which should be out soon - this is something that might be good to add to the leaderboard in the future (although I understand why it's not there atm, as it's very difficult to measure)

agreed. i've been thinking about that for the last month. i believe measuring inference speed could be done in a standardized way (looping over different system sizes and measuring time for single [and maybe multiple] forward pass[es]) which would allow plotting a 2D scatter plot of CPS / F1 / kappa / RMSD vs inference speed. i think this would be hugely useful as it establishes Pareto optimal models as a function of compute cost, allowing users to pick the optimal model based on their cost-accuracy requirements. if someone like NERSC is willing to provide an A100 or H100 for occasional use to test model speed, it would mostly be a question of automating the process as much as possible to lower maintenance burden.

the new metric vs model params plot is a first step in that direction as model params correlate with mem usage and to a much lesser degree with speed. for tools like [`torch-sim`](https://github.com/Radical-AI/torch-sim) that fill up GPU memory by batching structures as system sizes allow, number of model params actually correlate with speed a lot.

![Screenshot 2025-04-07 at 12 59 31 PM](https://github.com/user-attachments/assets/231f83e6-b697-45db-b2e1-5770e9b0f700)

DeNeutoy commented:
At least for the Orb models (and I would think most edge heavy NNPs), the 2 dominant things are actually not necessarily model parameter size, but:
- **edge featurization** - changing the number of edges has a massive impact in throughput, but also memory use - actually in the regime of relatively small models (e.g not LLM sized models), the main memory bottleneck is the activations from edge MLPs rather than model parameters themselves, particularly for conservative models which have to store 2x the number of activations
- **Conservatism**, as this requires 2x more memory by default (it is possible to get around this with checkpointing, but somewhat tricky to get it to play nicely with grads of grads for training)

One interesting thing also is the distinction between graph featurization and model forward timings as in some cases one or the other can be amortized by the workflow in question (e.g by caching a KNN graph). It's also important to look at scalability across atom sizes, because some approaches scale much worse (particularly for graph construction) than others. 

Once the paper is out I will clean up our benchmarking script that we used to measure things as it might be useful here - the basic idea is to have a interface which requires a calculator to separate out it's forward pass from graph construction, and then uses some torch benchmarking tooling to measure the different pieces.

DeNeutoy commented:
@janosh i've added the code for the thermal conductivity evaluation, sorry if there is some duplication - we pulled it out as we were experimenting with various things (e.g some precision experiments which will be in the paper). I've also added the displacements to the aws bucket:

```
aws s3 ls s3://orbitalmaterials-public-models/matbench-evals/orb-v3/
2025-04-08 12:32:12  108303375 orb-v3-con-inf-mpa-20250404-disp0.01-float32-highest-force-sets.json.gz
2025-04-08 12:31:33    8284016 orb-v3-con-inf-mpa-20250404-disp0.01-float32-highest-kappa.json.gz
2025-04-03 21:19:06  112263811 orb-v3-con-inf-mpa-20250404-disp0.03-float64-force-sets.json.gz
2025-04-03 21:18:24    8187377 orb-v3-con-inf-mpa-20250404-disp0.03-float64-kappa.json.gz
2025-04-08 12:32:54  114891530 orb-v3-con-inf-mpa-20250404-disp0.05-float32-highest-force-sets.json.gz
2025-04-08 12:33:17    8212033 orb-v3-con-inf-mpa-20250404-disp0.05-float32-highest-kappa.json.gz
2025-04-03 21:16:24    9006481 orb-v3-con-inf-mpa-20250404-symprec=1e-2-moyo=0.4.2.csv.gz
2025-04-03 21:16:36    6899746 orb-v3-con-inf-mpa-20250404-symprec=1e-5-moyo=0.4.2.csv.gz
2025-04-03 21:13:20    2748602 orb-v3-con-inf-mpa-20250404.csv.gz
2025-04-03 21:13:38  259885686 orb-v3-con-inf-mpa-20250404.json.gz
```

Results from these:
```
SRE/kSRME 0.01: 0.13352/0.21674
SRE/kSRME 0.05: 0.12172/0.20627
```

CompRhys commented on models/orb/orb-v3.yml:
```diff
@@ -0,0 +1,170 @@
+trained_for_benchmark: true
+train_task: S2EFS
+test_task: IS2RE-SR
+targets: EFS_D
```
Is this model conservative? EFS_D is EFS direct but the URL suggests it's conservative therefore EFS_G would be the right thing here

DeNeutoy replied:
Ah, yep! will fix

janosh commented:
thanks @DeNeutoy! 👍 

to avoid code duplication, I swaped `models/orb/thermal_conductivity/*.py` files for `test_orb_kappa.py` after confirming it gives identical results for the first 10 materials with below Cursor-generated code:


<details><summary>Compare kappa preds</summary>

```py
"""Compare force sets between two kappa.json.gz files and provide a simple example of self-comparison.

This script loads force data from two different kappa.json.gz files, matches materials
by their IDs, and compares both the mean forces and the full force arrays.
The script can be used directly with command line arguments or through its main function
which provides a self-comparison example.
"""

import argparse
import os
from typing import Any

import numpy as np
import pandas as pd
from tqdm import tqdm


def load_kappa_file(file_path: str) -> pd.DataFrame:
    """Load a kappa.json.gz file into a pandas DataFrame.

    Args:
        file_path: Path to the kappa.json.gz file

    Returns:
        DataFrame with material IDs as index

    Raises:
        FileNotFoundError: If the specified file cannot be found
        ValueError: If the file cannot be loaded as a JSON
    """
    if not os.path.isfile(file_path):
        raise FileNotFoundError(f"File not found: {file_path}")

    try:
        df = pd.read_json(file_path)
        # Set material_id as index if it exists as a column
        if "material_id" in df.columns:
            df = df.set_index("material_id")
        return df
    except Exception as exc:
        raise ValueError(f"Failed to load {file_path}: {exc}") from exc


def compare_force_sets(
    file1: str, file2: str, output_file: str | None = None, verbose: bool = False
) -> dict[str, dict[str, Any]]:
    """Compare force sets between two kappa.json.gz files.

    Args:
        file1: Path to the first kappa.json.gz file
        file2: Path to the second kappa.json.gz file
        output_file: Optional path to save results as CSV
        verbose: Whether to print detailed information during processing

    Returns:
        Dictionary with comparison results for each material ID
    """
    print(f"Loading file 1: {file1}")
    df1 = load_kappa_file(file1)

    print(f"Loading file 2: {file2}")
    df2 = load_kappa_file(file2)

    # Find common material IDs
    common_ids: set[str] = set(df1.index).intersection(set(df2.index))
    print(
        f"Found {len(common_ids)} common material IDs out of {len(df1)} in file1 and {len(df2)} in file2"
    )

    results: dict[str, dict[str, Any]] = {}

    # Check if 'fc2_set' and 'fc3_set' are directly in the dataframes
    force_columns: list[str] = []
    if "fc2_set" in df1.columns:
        force_columns.append("fc2_set")
    if "fc3_set" in df1.columns:
        force_columns.append("fc3_set")

    # If no force columns found, we might need to check for a separate force-sets.json.gz file
    if not force_columns:
        print(
            "No force columns found. Please ensure the files contain force data or provide force-sets.json.gz files."
        )
        return results

    print(f"Comparing {', '.join(force_columns)} for each material...")
    for mat_id in tqdm(common_ids):
        results[mat_id] = {}

        for force_col in force_columns:
            # Get force arrays for this material
            forces1 = df1.loc[mat_id, force_col]
            forces2 = df2.loc[mat_id, force_col]

            # Skip if forces not available
            if forces1 is None or forces2 is None:
                results[mat_id][f"{force_col}_available"] = False
                continue

            # Convert to numpy arrays if they aren't already
            forces1_array = np.array(forces1)
            forces2_array = np.array(forces2)

            # Check if shapes match
            shape_match = forces1_array.shape == forces2_array.shape
            results[mat_id][f"{force_col}_shape_match"] = shape_match

            if not shape_match:
                results[mat_id][f"{force_col}_shape1"] = forces1_array.shape
                results[mat_id][f"{force_col}_shape2"] = forces2_array.shape
                continue

            # Calculate means and differences
            mean1 = np.mean(forces1_array)
            mean2 = np.mean(forces2_array)
            mean_diff = mean1 - mean2

            # Calculate absolute differences
            abs_diff = np.abs(forces1_array - forces2_array)
            max_abs_diff = np.max(abs_diff)
            mean_abs_diff = np.mean(abs_diff)

            # Calculate relative differences where forces aren't near zero
            # Avoid division by zero or near-zero values
            mask = np.abs(forces2_array) > 1e-10
            if np.any(mask):
                rel_diff = np.abs(
                    (forces1_array[mask] - forces2_array[mask]) / forces2_array[mask]
                )
                max_rel_diff = np.max(rel_diff)
                mean_rel_diff = np.mean(rel_diff)
            else:
                max_rel_diff = np.nan
                mean_rel_diff = np.nan

            # Store results
            results[mat_id][f"{force_col}_mean1"] = float(mean1)
            results[mat_id][f"{force_col}_mean2"] = float(mean2)
            results[mat_id][f"{force_col}_mean_diff"] = float(mean_diff)
            results[mat_id][f"{force_col}_max_abs_diff"] = float(max_abs_diff)
            results[mat_id][f"{force_col}_mean_abs_diff"] = float(mean_abs_diff)
            results[mat_id][f"{force_col}_max_rel_diff"] = float(max_rel_diff)
            results[mat_id][f"{force_col}_mean_rel_diff"] = float(mean_rel_diff)

            if verbose:
                print(f"\nMaterial {mat_id} - {force_col}:")
                print(f"  Mean force 1: {mean1:.6e}")
                print(f"  Mean force 2: {mean2:.6e}")
                print(f"  Mean difference: {mean_diff:.6e}")
                print(f"  Max absolute difference: {max_abs_diff:.6e}")
                print(f"  Mean absolute difference: {mean_abs_diff:.6e}")
                print(f"  Max relative difference: {max_rel_diff:.6e}")
                print(f"  Mean relative difference: {mean_rel_diff:.6e}")

    # Calculate summary statistics
    summary = calculate_summary_statistics(results, force_columns)

    # Save results to file if requested
    if output_file:
        results_df = pd.DataFrame.from_dict(results, orient="index")
        results_df.to_csv(output_file)
        print(f"Results saved to {output_file}")

    # Print summary
    print("\nSummary Statistics:")
    for force_col in force_columns:
        print(f"\n{force_col}:")
        for key, value in summary[force_col].items():
            print(f"  {key}: {value}")

    return results


def calculate_summary_statistics(
    results: dict[str, dict[str, Any]], force_columns: list[str]
) -> dict[str, dict[str, float]]:
    """Calculate summary statistics from the comparison results.

    Args:
        results: Dictionary containing comparison results for each material
        force_columns: List of force column names

    Returns:
        Dictionary with summary statistics for each force column
    """
    summary: dict[str, dict[str, float]] = {}

    for force_col in force_columns:
        col_summary: dict[str, float] = {}

        # Collect all values for this force column
        mean_diffs: list[float] = []
        abs_diffs: list[float] = []
        rel_diffs: list[float] = []

        for mat_id, mat_results in results.items():
            if f"{force_col}_mean_diff" in mat_results:
                mean_diffs.append(mat_results[f"{force_col}_mean_diff"])
            if f"{force_col}_mean_abs_diff" in mat_results:
                abs_diffs.append(mat_results[f"{force_col}_mean_abs_diff"])
            if f"{force_col}_mean_rel_diff" in mat_results:
                rel_diffs.append(mat_results[f"{force_col}_mean_rel_diff"])

        # Calculate statistics
        col_summary["mean_difference_avg"] = (
            np.mean(mean_diffs) if mean_diffs else np.nan
        )
        col_summary["mean_difference_std"] = (
            np.std(mean_diffs) if mean_diffs else np.nan
        )
        col_summary["abs_difference_avg"] = np.mean(abs_diffs) if abs_diffs else np.nan
        col_summary["abs_difference_std"] = np.std(abs_diffs) if abs_diffs else np.nan
        col_summary["rel_difference_avg"] = np.mean(rel_diffs) if rel_diffs else np.nan
        col_summary["rel_difference_std"] = np.std(rel_diffs) if rel_diffs else np.nan
        col_summary["num_materials_compared"] = len(mean_diffs)

        summary[force_col] = col_summary

    return summary


def run_self_comparison_example() -> None:
    """Run a comparison of a force-sets.json.gz file with itself as an example.

    This is a simple demonstration of how to use the compare_force_sets function
    with actual files in the project.
    """
    # Path to the force sets file
    file_path = "./kappa_results_orb-v3-conservative-inf-mpa/force-sets.json.gz"

    # Ensure the file exists
    if not os.path.isfile(file_path):
        print(f"Error: Could not find {file_path}")
        print(
            "Please make sure you're running this script from the project root directory."
        )
        return

    # Compare the file with itself (should show minimal differences)
    print("Comparing force sets file with itself (self-comparison test)...")
    compare_force_sets(
        file1=file_path,
        file2=file_path,
        output_file="force_sets_comparison_results.csv",
        verbose=True,
    )

    print(
        "\nIf you want to compare different files, use this script with command line arguments:"
    )
    print("python compare_force_sets.py <file1> <file2> -o results.csv -v")


def main() -> None:
    """Parse command line arguments and run the comparison or example if no args provided."""
    parser = argparse.ArgumentParser(
        description="Compare force sets between two kappa.json.gz files"
    )
    parser.add_argument(
        "file1",
        type=str,
        nargs="?",
        help="Path to the first kappa.json.gz file or force-sets.json.gz file",
    )
    parser.add_argument(
        "file2",
        type=str,
        nargs="?",
        help="Path to the second kappa.json.gz file or force-sets.json.gz file",
    )
    parser.add_argument(
        "-o",
        "--output",
        type=str,
        help="Path to save detailed results as CSV",
    )
    parser.add_argument(
        "-v",
        "--verbose",
        action="store_true",
        help="Print detailed information during processing",
    )
    parser.add_argument(
        "--example",
        action="store_true",
        help="Run the self-comparison example",
    )

    args = parser.parse_args()

    # If example flag is provided or no arguments given, run the example
    if args.example or (args.file1 is None and args.file2 is None):
        run_self_comparison_example()
    elif args.file1 and args.file2:
        compare_force_sets(args.file1, args.file2, args.output, args.verbose)
    else:
        parser.print_help()


if __name__ == "__main__":
    main()

```
</details> 

I also deleted `models/orb/analyse_geo_opt.py` after confirming I get the same results with `python scripts/analyze_geo_opt.py --models orb_v3`.
